File: src/main/java/org/elmlang/intellijplugin/features/structure/ElmStructureViewFactory.java
package org.elmlang.intellijplugin.features.structure;

import com.intellij.ide.structureView.StructureViewBuilder;
import com.intellij.ide.structureView.StructureViewModel;
import com.intellij.ide.structureView.TreeBasedStructureViewBuilder;
import com.intellij.lang.PsiStructureViewFactory;
import com.intellij.openapi.editor.Editor;
import com.intellij.psi.PsiFile;
import org.elmlang.intellijplugin.psi.ElmFile;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class ElmStructureViewFactory implements PsiStructureViewFactory {
    @Nullable
    @Override
    public StructureViewBuilder getStructureViewBuilder(PsiFile psiFile) {
        ElmFile elmFile = (ElmFile) psiFile;
        return new TreeBasedStructureViewBuilder() {
            @NotNull
            @Override
            public StructureViewModel createStructureViewModel(@Nullable Editor editor) {
                return new ElmStructureViewModel(elmFile);
            }
        };
    }
}

File: src/main/java/org/elmlang/intellijplugin/features/structure/ElmStructureViewFactory.kt
package org.elmlang.intellijplugin.features.structure

import com.intellij.ide.structureView.TreeBasedStructureViewBuilder
import com.intellij.lang.PsiStructureViewFactory
import com.intellij.openapi.editor.Editor
import com.intellij.psi.PsiFile
import org.elmlang.intellijplugin.psi.ElmFile

class ElmStructureViewFactory: PsiStructureViewFactory {

    override fun getStructureViewBuilder(psiFile: PsiFile?): TreeBasedStructureViewBuilder {
        val elmFile = psiFile as ElmFile
        return object : TreeBasedStructureViewBuilder() {
            override fun createStructureViewModel(editor: Editor?) =
                    ElmStructureViewModel(elmFile)
        }
    }
}